### PR TITLE
Add Role Mapping Cleanup Security Migration

### PR DESCRIPTION
**Pull request by jfreden**
This PR adds a cleanup security migration for role mapping duplicates that exist both in cluster state and the `.security` index. The cleanup task is implemented as a [security migration](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java). 

The task compares the names of the role mappings in the operator-defined `settings.json` and the names of the role mappings in the `.security` index and removes the ones in the `.security` index. 

## Issue
This cleanup is done because there was a bug (fixed [here](https://github.com/elastic/elasticsearch/pull/114337)) that caused ECK customers to have the same role mappings with the same names present both in cluster state and the `.security` index. The effective role mapping is the combination of the `.security` index mapping and the cluster state one. After the bug [fix](https://github.com/elastic/elasticsearch/pull/114337), if the mapping in cluster state is modified (through `setting.json`), the `.security` index will still contain the old mapping and this could lead to unintended behaviour.

## Dependencies
- https://github.com/elastic/elasticsearch/pull/114295
- This migration doesn't have dependencies on prior migrations, but since we're adding more migrations, I've also created this PR as dependency/nice to have before merging this: https://github.com/elastic/elasticsearch/pull/114879

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/IndexVersions.java [hunk 1]
```diff
@@ -117,6 +117,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion ENABLE_IGNORE_MALFORMED_LOGSDB = def(8_514_00_0, Version.LUCENE_9_11_1);
     public static final IndexVersion MERGE_ON_RECOVERY_VERSION = def(8_515_00_0, Version.LUCENE_9_11_1);
     public static final IndexVersion UPGRADE_TO_LUCENE_9_12 = def(8_516_00_0, Version.LUCENE_9_12_0);
+    public static final IndexVersion ADD_ROLE_MAPPING_MIGRATION = def(8_517_00_0, Version.LUCENE_9_12_0);
```
Comment (jfreden) [inline server/src/main/java/org/elasticsearch/index/IndexVersions.java:120]: This is needed to allow us to check if an index is newly created or not. If an index is new we can shortcut the migration and skip it (nothing to migrate).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/ReservedRoleMappingAction.java [hunk 2]
```diff
@@ -34,6 +35,8 @@
  */
 public class ReservedRoleMappingAction implements ReservedClusterStateHandler<List<PutRoleMappingRequest>> {
     public static final String NAME = "role_mappings";
+    private static final String FILE_SETTINGS_METADATA_NAMESPACE = "file_settings";
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/ReservedRoleMappingAction.java:38]: Consider moving these. These constants are not available in the security plugin so need to be defined somewhere.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java [hunk 3]
```diff
@@ -100,6 +109,21 @@ private void applyOutstandingMigrations(AllocatedPersistentTask task, int curren
         }
     }
 
+    /**
+     * Refresh security index to make sure that docs that were migrated are visible to the next migration and to prevent version conflicts
+     * or unexpected behaviour by APIs relying on migrated docs.
+     */
+    private void refreshSecurityIndex(ActionListener<Void> listener) {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java:116]: This change was suggested by Albert and I think it would be nice to get in now that we have a second migration. See [this](https://github.com/elastic/elasticsearch/issues/110562) issue. It could be a separate PR.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 4]
```diff
@@ -267,6 +272,11 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
         return indexVersionCreated != null && indexVersionCreated.onOrAfter(IndexVersion.current());
     }
 
+    private static Set<String> getFileSettingsMetadataHandlerRoleMappingKeys(ClusterState clusterState) {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:275]: This isn't exactly part of the security index, so this doesn't belong here. For simplicity I added it here. Ideally the security migration state should be tracked in a separate cluster state handler. I think that's an improvement/refactor that should be done in a separate PR if we keep adding migrations outside the security index.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            assumeTrue(
+                "Cleanup only needed before security.role_mapping_cleanup feature available in cluster",
+                clusterHasFeature("security.role_mapping_cleanup") == false
+            );
+            // Since the old cluster has role mappings in cluster state, but doesn't check duplicates, create a duplicate
+            createNativeRoleMapping("operator_role_mapping_1", Map.of("meta", "test"));
+            createNativeRoleMapping("no_name_conflict", Map.of("meta", "test"));
+            assertAllRoleMappings(client(), "no_name_conflict", "operator_role_mapping_1");
+        } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
+            waitForSecurityMigrationCompletion(adminClient(), 2);
```
Comment (jfreden) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:46]: These tests will fail here until the file service change is merged, since the preconditions for the migration are not met.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ExpressionRoleMapping.java [hunk 6]
```diff
@@ -189,7 +189,7 @@ public List<TemplateRoleName> getRoleTemplates() {
      * This is not used within the mapping process, and does not affect whether the expression matches, nor which roles are assigned.
      */
     public Map<String, Object> getMetadata() {
-        return Collections.unmodifiableMap(metadata);
+        return metadata != null ? Collections.unmodifiableMap(metadata) : Collections.emptyMap();
```
Comment (jfreden) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/mapper/ExpressionRoleMapping.java:192]: These are not set to null anywhere in production code, but could be in the future (since constructor is public) and can in tests (as I found out).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 7]
```diff
@@ -62,63 +85,159 @@ public interface SecurityMigration {
...
-                        updateRolesByQuery(indexManager, client, filterQuery, listener);
+        Map.of(
+            ROLE_METADATA_FLATTENED_MIGRATION_VERSION,
+            new RoleMetadataFlattenedMigration(),
+            CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION,
+            new CleanupRoleMappingDuplicatesMigration()
+        )
+    );
+
+    public static class RoleMetadataFlattenedMigration implements SecurityMigration {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:100]: No changes here, just moved this to a named class since I wanted `CleanupRoleMappingDuplicatesMigration` to be a named class for testing purposes and it felt weird to have one anonymous migration and one that is not.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 8]
```diff
@@ -62,63 +85,159 @@ public interface SecurityMigration {
...
+            return Set.of(SecuritySystemIndices.SECURITY_ROLES_METADATA_FLATTENED);
+        }
+
+        @Override
+        public int minMappingVersion() {
+            return ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS.id();
+        }
+    }
+
+    public static class CleanupRoleMappingDuplicatesMigration implements SecurityMigration {
```
Comment (jfreden) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:150]: The new migration

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 9]
```diff
@@ -267,6 +270,23 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
         return indexVersionCreated != null && indexVersionCreated.onOrAfter(IndexVersion.current());
     }
 
+    /**
+     * Check to see if there are any file based role mappings and if they have been loaded into cluster state. If the
+     * {@link ReservedStateMetadata} file_settings namespace contains role mapping names, it means that there should be the same number of
+     * cluster state role mappings available. If they're not available yet using {@code RoleMappingMetadata.getFromClusterState()}, they
+     * have not yet been synchronized.
+     */
+    private static boolean isReservedRoleMappingsSynced(ClusterState clusterState) {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:279]: WDYT about calling this `readyForRoleMappingCleanupMigration`? It feels more relevant as a security-index concept. 

Alternatively, I _think_ we could entirely remove this from the index manager if we expose a `client` instance (or the `ClusterService`) to the `checkPreConditions` method. That would give us cluster state instead of relying on the index manager.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 10]
```diff
@@ -267,6 +270,23 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
...
+        ReservedStateMetadata fileSettingsMetadata = clusterState.metadata().reservedStateMetadata().get(FILE_SETTINGS_METADATA_NAMESPACE);
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:280]: I'd move the bit where we're retrieving handler metadata to `ReservedStateMetadata`, something like:

```java
public Set<String> keys(String handlerName) {
  ReservedStateHandlerMetadata handlerMetadata = handlers.get(handlerName);
  if (handlerMetadata == null || handlerMetadata.keys().isEmpty()) {
      return Collections.emptySet();
  }

  return Collections.unmodifiableSet(handlerMetadata.keys());
}
```

This would allow us to keep some of the logic of retrieving keys encapsulated.

I'd take it a step further and remove the `get(FILE_SETTINGS_METADATA_NAMESPACE)`, and instead just assert that we have 0 or 1 namespaces. We rely on this assumption in many other places, and I think as long as we fail very openly it will get caught by CI, should we ever add another namespace. 

Basically, just want to avoid duplicating string constants across modules. So something like:

```java
Map<String, ReservedStateMetadata> reservedStateMetadataMap = clusterState.metadata().reservedStateMetadata();
assert reservedStateMetadataMap.values().size() <= 1;
for (ReservedStateMetadata reservedStateMetadata : reservedStateMetadataMap.values()) {
    var keys = reservedStateMetadata.keys(ReservedRoleMappingAction.NAME);
    if (keys.isEmpty()) {
        var other = RoleMappingMetadata.getFromClusterState(clusterState).getRoleMappings().size();
        return other == keys.size();
    }
}
return true;
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 11]
```diff
@@ -62,63 +85,159 @@ public interface SecurityMigration {
...
+                ActionListener.wrap(response -> listener.onResponse(response.mappings()), listener::onFailure)
+            );
+        }
+
+        private void deleteNativeRoleMappings(Client client, Iterator<String> namesIterator, ActionListener<Void> listener) {
+            String name = namesIterator.next();
+            executeAsyncWithOrigin(
+                client,
+                SECURITY_ORIGIN,
+                DeleteRoleMappingAction.INSTANCE,
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:179]: WDYT about a bulk-delete against the raw index instead? Using the API is nice but it's not one-shot where if a single delete OP fails the whole migration stops (at least based on the current implementation). I haven't checked yet but the Header warnings the delete API returns due to name-clash might get logged too, which could confuse users.

**Review by n1v0lg — COMMENTED** [review]
Looking good! Doing this via security migration feels like the way to go 👍 

Leaving some initial thoughts, will continue the review tomorrow!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java (hunk 11 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:179]: Yes, that would be ideal and I actually went down that route first, since it's really nice to be able to do it in one shot. 

The reason I switched to use the action was the cache clearing [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java#L215). Just updating the index won't clear the cache so we'd have to add a step to that too in the migration. 

Adding a step to clear the cache might be a good option, but I opted for doing it one by one instead to make sure we use a well tested path (it's another thing that could fail). 

>I haven't checked yet but the Header warnings the delete API returns due to name-clash might get logged too, which could confuse users.

Since we use the transport action directly (no http headers), they should not get logged. I can do some testing to confirm.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java (hunk 9 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:279]: > WDYT about calling this readyForRoleMappingCleanupMigration

My intention was to decouple it from the migration but I guess that's pretty silly, because it's only ever going to be used for that. I'll update it to be more explicit. 

>Alternatively, I think we could entirely remove this from the index manager if we expose a client instance (or the ClusterService) to the checkPreConditions method. That would give us cluster state instead of relying on the index manager.

The migrations are only triggered on security index manager state changes (it's a listener). So if the migration is not ready to be executed, we need a way to make sure that the state changes when it is and that's why this needs to be part of the state. 

That's what I'm referring to here: https://github.com/elastic/elasticsearch/pull/114830#discussion_r1801226195 ideally I think there should be a separate state that track the migrations, where we can track state changes that are not necessarily specific to the security index.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java (hunk 10 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:280]: I added your suggested `keys` method. That made it a lot cleaner! Thank you! 

I've update the code to use the `ReservedRoleMappingAction.NAME` instead of the new unnecessary constant I added.  

I hear you with duplicating strings, but I find the assumption and loop confusing. I think I rather keep the duplicated string for readability, but I don't feel super strongly.

Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:280]: 👍 if we duplicate the string, could you still add an assertion that ensures that if we get back nothing for the string, we don't have _any_ namespaces. I want to avoid us adding a namespace in the future/changing the original string/other unrelated changes to go unnoticed in the migration code

Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:280]: Yes! Good idea.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java (hunk 11 above)
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:179]: [This class](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/AbstractRoleMapperClearRealmCache.java) is responsible for cache clearing and all it does is call a transport action with the realm names. 

We could do that and skip the names since that means a clear all realms. Looks like we also need to clear the local cache realms. All of these things require pretty rigorous testing and it's easy to mess up. 

We've decided to instead make sure that if a delete fails, the following ones are executed, to make sure one failed delete doesn't cause all of them to fail.

File: server/src/main/java/org/elasticsearch/index/IndexVersions.java [hunk 12]
```diff
@@ -117,6 +117,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion ENABLE_IGNORE_MALFORMED_LOGSDB = def(8_514_00_0, Version.LUCENE_9_11_1);
     public static final IndexVersion MERGE_ON_RECOVERY_VERSION = def(8_515_00_0, Version.LUCENE_9_11_1);
     public static final IndexVersion UPGRADE_TO_LUCENE_9_12 = def(8_516_00_0, Version.LUCENE_9_12_0);
+    public static final IndexVersion ADD_ROLE_MAPPING_CLEANUP_MIGRATION = def(8_517_00_0, Version.LUCENE_9_12_0);
```
[the remark this answers is not in the document]
Comment (jfreden) [reply server/src/main/java/org/elasticsearch/index/IndexVersions.java:120]: Good point! We have to be careful when creating the backport PRs.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 13]
```diff
@@ -62,63 +85,161 @@ public interface SecurityMigration {
...
+        }
+
+        @Override
+        public Set<NodeFeature> nodeFeaturesRequired() {
+            return Set.of(SecuritySystemIndices.SECURITY_ROLE_MAPPING_CLEANUP);
+        }
+
+        @Override
+        public int minMappingVersion() {
+            return ADD_MANAGE_ROLES_PRIVILEGE.id();
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:218]: The idea is that this should be referring to the mapping version that the migration was created on. It doesn't have to be updated if no mapping changes happened. I realize now that this is a little confusing and I can add a follow up PR to try to address this.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:218]: Got it, thanks for explaining! Yup, lets just a comment to that effect then.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 14]
```diff
@@ -62,63 +85,161 @@ public interface SecurityMigration {
...
+            executeAsyncWithOrigin(
+                client,
+                SECURITY_ORIGIN,
+                GetRoleMappingsAction.INSTANCE,
+                new GetRoleMappingsRequestBuilder(client).request(),
+                ActionListener.wrap(response -> listener.onResponse(response.mappings()), listener::onFailure)
+            );
+        }
+
+        private void deleteNativeRoleMappings(Client client, Iterator<String> namesIterator, ActionListener<Void> listener) {
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java]: Updated to use the `GroupedActionListener`. Thanks for that suggestion!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 15]
```diff
@@ -267,6 +270,21 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
...
+     * Check to see if there are any file based role mappings and if they have been loaded into cluster state. If the
+     * {@link ReservedStateMetadata} file_settings namespace contains role mapping names, it means that there should be the same number of
+     * cluster state role mappings available. If they're not available yet using {@code RoleMappingMetadata.getFromClusterState()}, they
+     * have not yet been synchronized.
+     */
+    private static boolean isReadyForRoleMappingCleanupMigration(ClusterState clusterState) {
+        ReservedStateMetadata fileSettingsMetadata = clusterState.metadata().reservedStateMetadata().get(FILE_SETTINGS_METADATA_NAMESPACE);
+        assert fileSettingsMetadata != null || clusterState.metadata().reservedStateMetadata().isEmpty();
+        return fileSettingsMetadata == null
+            || RoleMappingMetadata.getFromClusterState(clusterState).getRoleMappings().size() == fileSettingsMetadata.keys(
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java]: Yes, added that. Good catch!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 16]
```diff
@@ -267,6 +270,21 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
         return indexVersionCreated != null && indexVersionCreated.onOrAfter(IndexVersion.current());
     }
 
+    /**
+     * Check to see if there are any file based role mappings and if they have been loaded into cluster state. If the
+     * {@link ReservedStateMetadata} file_settings namespace contains role mapping names, it means that there should be the same number of
+     * cluster state role mappings available. If they're not available yet using {@code RoleMappingMetadata.getFromClusterState()}, they
+     * have not yet been synchronized.
+     */
+    private static boolean isReadyForRoleMappingCleanupMigration(ClusterState clusterState) {
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java]: Great idea! I actually started with an enum, then when we had the cluster mappings available in the security index manager state, I removed it (since we could just check size == 0 and return in the migration). Now we've come full circle and I've added it back :) 

I've also added a status `DONE` so we don't have to do these calculations on every security index manager change forever, which is really nice.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:225]: Created this issue for the improvement: https://github.com/elastic/elasticsearch/issues/115497

**Review by n1v0lg — COMMENTED** [review]
Prod code looks solid 💯 I'll take another pass over it Monday and go through tests as well.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 17]
```diff
@@ -267,6 +277,49 @@ private static boolean isCreatedOnLatestVersion(IndexMetadata indexMetadata) {
...
+     * migration
+     *
+     * @param clusterState current cluster state
+     * @param migrationsVersion current migration version
+     *
+     * @return RoleMappingsCleanupMigrationStatus
+     */
+    static RoleMappingsCleanupMigrationStatus getRoleMappingsCleanupMigrationStatus(ClusterState clusterState, int migrationsVersion) {
+        // Migration already finished
+        if (migrationsVersion >= SecurityMigrations.CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION) {
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:291]: Correct. `migrationsVersion` comes from cluster state and will only be incremented when all migrations prior to the current version has been successfully applied in order. The value of the constant `CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION` is never intended to be updated.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 18]
```diff
@@ -62,63 +88,165 @@ public interface SecurityMigration {
...
+        public int minMappingVersion() {
+            return ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS.id();
+        }
+    }
+
+    public static class CleanupRoleMappingDuplicatesMigration implements SecurityMigration {
+        @Override
+        public void migrate(SecurityIndexManager indexManager, Client client, ActionListener<Void> listener) {
+            if (indexManager.getRoleMappingsCleanupMigrationStatus() == SecurityIndexManager.RoleMappingsCleanupMigrationStatus.SKIP) {
+                listener.onResponse(null);
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:158]: I was debating this with myself and landed on not adding one here. 

This migration will only run for ECK customers with operator defined role mappings, all others will hit this condition. I feel hesitant to adding a log that nothing was done that will almost always be true. We already have an [info log](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java#L98) that says what migration version was applied and a [log line ](https://github.com/elastic/elasticsearch/pull/114830/files#diff-4393c8837a34bb39a3e289d942366215191c7c52c430bbdffa0c9829d51a2beaR166) that logs what will be cleaned up. I don't think we should add another one.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java [hunk 19]
```diff
@@ -62,63 +88,165 @@ public interface SecurityMigration {
...
+                ActionListener.wrap(responses -> {
+                    Map<Boolean, List<DeleteRoleMappingResponse>> deletesPartitionedByFound = responses.stream()
+                        .collect(Collectors.partitioningBy(DeleteRoleMappingResponse::isFound));
+                    if (deletesPartitionedByFound.containsKey(false)) {
+                        logger.warn(
+                            "[" + deletesPartitionedByFound.get(false).size() + "] Role mappings not found during role mapping clean up."
+                        );
+                    }
+                    if (deletesPartitionedByFound.containsKey(true)) {
+                        logger.info("Deleted [" + deletesPartitionedByFound.get(true) + "] duplicated role mapping from .security index");
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java:196]: Yes, sorry about this logging disaster. I've simplified it and dropped the partitioned by thing.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.TransportVersions.SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+public class SecurityIndexRoleMappingCleanupIT extends AbstractUpgradeTestCase {
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:26]: I ended up changing this to `contains` and I found a bug in the old file handling. I think we might want to create a patch and backport to versions before 8.15.

The migration never executed because there wasn't a security index (expected behaviour from the migration), but since the role mappings should have been created in the security index it was strange that the index didn't exist. Turns out the call to the native role mapping store [here](https://github.com/elastic/elasticsearch/blob/8.14/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/ReservedRoleMappingAction.java#L129) happens while the `SecurityIndexManager` is in [unrecovered state](https://github.com/elastic/elasticsearch/blob/8.14/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java#L552). This leads to the native role mapping store thinking that the index is not up to date [here](https://github.com/elastic/elasticsearch/blob/8.14/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java#L204) and nothing is inserted in to the store.

This won't impact the current state of things, so we can proceed, but we should discuss how to address this after this fix.

File: server/src/main/java/org/elasticsearch/index/IndexVersions.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (jfreden) [reply server/src/main/java/org/elasticsearch/index/IndexVersions.java]: After discussing with core infra, I opted to go for an index setting, see this commit: https://github.com/elastic/elasticsearch/pull/114830/commits/9533f6f331e937abfacd19c2c929a7764f233549 

Did a small POC in this PR to test this before: https://github.com/elastic/elasticsearch/pull/115666

The benefit with this a approach is that we no longer need to manually bump index versions when adding new migrations and that it's not as heavy weight since we no longer need to bump the index version for all indices.

**Review by n1v0lg — COMMENTED** [review]
Tests look great, too. A few coverage suggestions -- I'll make one more pass when all is ready. Thanks for iterating on this!

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/support/CleanupRoleMappingDuplicatesMigrationIT.java [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+        final String masterNode = internalCluster().getMasterName();
+        ensureGreen();
+        CountDownLatch awaitMigrations = awaitMigrationVersionUpdates(
+            masterNode,
+            SecurityMigrations.CLEANUP_ROLE_MAPPING_DUPLICATES_MIGRATION_VERSION
+        );
+        // Create a native role mapping to create security index and trigger migration
+        createNativeRoleMapping("everyone_kibana_alone");
+        // Make sure no migration ran (set to current version without applying prior migrations)
+        safeAwait(awaitMigrations);
```
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/support/CleanupRoleMappingDuplicatesMigrationIT.java:275]: Do you mean: 
1. Create security index
2. Validate that migration was skipped
3. Delete security index 
4. Create security index again
5. Validate that migration was skipped

It feels like that's testing the same thing twice.

**Comment by jfreden** [comment]
@n1v0lg Thanks for the review! I've addressed all the comments and reverted the index settings like we discussed offline.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/support/CleanupRoleMappingDuplicatesMigrationIT.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/support/CleanupRoleMappingDuplicatesMigrationIT.java:315]: yeah... written out that does indeed feel redundant. lets skip 👍

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+                minimumTransportVersion().after(SECURITY_ROLE_MAPPINGS_IN_CLUSTER_STATE)
+            );
+            // Since the old cluster has role mappings in cluster state, but doesn't check duplicates, create duplicates
+            createNativeRoleMapping("operator_role_mapping_1", Map.of("meta", "test"));
+            createNativeRoleMapping("operator_role_mapping_2", Map.of("meta", "test"));
+            // API will return only native role mappings
+            assertAllRoleMappings(client(), "operator_role_mapping_1", "operator_role_mapping_2");
+        } else if (CLUSTER_TYPE == ClusterType.MIXED) {
+            // Create a native role mapping that doesn't conflict with anything before the migration run
+            createNativeRoleMapping("no_name_conflict", Map.of("meta", "test"));
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:48]: Ah whoops, my bad -- in a mixed cluster `createNativeRoleMapping("native_role_mapping" + ExpressionRoleMapping.READ_ONLY_ROLE_MAPPING_SUFFIX, Map.of("meta", "test"));` can result in a rejected request -- I think we want to move this to the `(CLUSTER_TYPE == ClusterType.OLD)` branch

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:26]: I think this may have actually gotten fixed here: https://github.com/elastic/elasticsearch/pull/92173

[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:26]: Good find!

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (jfreden) [reply x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SecurityIndexRoleMappingCleanupIT.java:48]: Ahh that's right! 

We can't create it there since we won't be bale to validate it. Not all bwc tests will hit the old cluster code because of the `assumeTrue` statements. I think we should skip it in this case for that reason.
